src/instrumentation/podmutator.go

```diff
@@ -18,111 +18,126 @@ package instrumentation
+}
+
+type instPodMutator struct {
+	client                        client.Client
+	sdkInjector                   SdkInjector
+	secretReplicator              SecretReplicator
+	instrumentationLocator        InstrumentationLocator
+	getLicenseKeySecretFn         func(insts []*v1alpha1.Instrumentation) (string, error)
+	getLanguageInstrumentationsFn func(insts []*v1alpha1.Instrumentation) (LanguageInstrumentations, error)
+	logger                        logr.Logger
+	operatorNamespace             string
+}
 
-func NewMutator(logger logr.Logger, client client.Client) *instPodMutator {
+// NewMutator is used to get a new instance of a mutator
+func NewMutator(
+	logger logr.Logger,
+	client client.Client,
+	sdkInjector SdkInjector,
+	secretReplicator SecretReplicator,
+	instrumentationLocator InstrumentationLocator,
+	getLicenseKeySecretFn func(insts []*v1alpha1.Instrumentation) (string, error),
+	getLanguageInstrumentationsFn func(insts []*v1alpha1.Instrumentation) (LanguageInstrumentations, error),
+	operatorNamespace string,
+) *instPodMutator {
 	return &instPodMutator{
-		Logger: logger,
-		Client: client,
-		sdkInjector: &sdkInjector{
-			logger: logger,
-			client: client,
-		},
+		logger:                        logger,
+		client:                        client,
+		instrumentationLocator:        instrumentationLocator,
+		sdkInjector:                   sdkInjector,
+		secretReplicator:              secretReplicator,
+		getLicenseKeySecretFn:         getLicenseKeySecretFn,
+		getLanguageInstrumentationsFn: getLanguageInstrumentationsFn,
+		operatorNamespace:             operatorNamespace,
 	}
 }
 
+// Mutate is used to mutate a pod based on some instrumentation(s)
 func (pm *instPodMutator) Mutate(ctx context.Context, ns corev1.Namespace, pod corev1.Pod) (corev1.Pod, error) {
-	logger := pm.Logger.WithValues("namespace", pod.Namespace, "name", pod.Name)
-
-	var inst *v1alpha1.Instrumentation
-	var err error
-
-	insts := languageInstrumentations{}
-
-	// We bail out if any annotation fails to process.
-
-	if inst, err = pm.getInstrumentationInstance(ctx, ns, pod, annotationInjectJava); err != nil {
-		// we still allow the pod to be created, but we log a message to the operator's logs
-		logger.Error(err, "failed to select a New Relic Instrumentation instance for this pod")
-		return pod, err
-	}
-	insts.Java = inst
-
-	if inst, err = pm.getInstrumentationInstance(ctx, ns, pod, annotationInjectNodeJS); err != nil {
-		// we still allow the pod to be created, but we log a message to the operator's logs
-		logger.Error(err, "failed to select a New Relic Instrumentation instance for this pod")
-		return pod, err
+	switch mode {
+	case "annotations":
+		return pm.MutateViaAnnotations(ctx, ns, pod)
+	case "poc", "first-matching-instrumentation":
+		return pm.MutateWithFirstInstrumentation(ctx, ns, pod)
+	case "last-matching-instrumentation-by-lang":
+		return pm.MutateWithAnyAvailInstrumentations(ctx, ns, pod)
```

Comment:
Let's set a `default` case here and make the default case as the `last-matching-instrumentation-by-lang`.

Reply:
That code was refactored